### ISSUE TITLE
Segmentation fault for larger grids in C++

### ISSUE THREAD

**Issue by diregoblin** [multiscale/muscle3#62]
Grids larger than a certain size (seems to be a few thousand elements) cause a segmentation fault when accessed through ``elements<>()`` after being received.

How to reproduce:

Module a: 
```
size_t t_cur = 0;
std::vector<int64_t> testVector (1000000);
auto test_result = Data::grid(testVector.data(), {100, 100, 100}, {"x", "y", "z"});
instance.send("gridPort", Message(t_cur, test_result));
```
Module b:
```
    auto domainObs = instance.receive("gridExit");
    size_t t_cur = domainObs.timestamp();
    std::cout << "Domain received for time " << t_cur << std::endl;

    size_t doSize = domainObs.data().size();
    std::vector<size_t> boxSize = domainObs.data().shape();
    std::cout << "Domain size " << doSize << " elements, " << boxSize[0] << " by " << boxSize[1] << " by " << boxSize[2] << std::endl;

    auto domainObsData = domainObs.data();
    if (domainObsData.is_a_grid_of<int64_t>())
        std::cout << "Received a grid" << std::endl;
    else
        std::cout << "Received a non-grid" << std::endl;

    auto domainObsPtr = domainObs.data().elements<int64_t>();
    std::cout << "Grid address  " << domainObsPtr << std::endl;
    std::vector<int64_t> test (domainObsPtr, domainObsPtr + doSize);
    std::cout << "test vector assigned" << std::endl;
```
For large grids, this code should crash with a segmentation fault when trying to assign the ``test`` vector.

**Comment by LourensVeen** [comment]
This seems to be a bug in `libmuscle`, it looks like the `elements()` function returns a pointer to data that has been deallocated. That's wrong for small grids as well, but I guess the segfault is only triggered for larger data where we actually go beyond the segment. Ouch!

The desired behaviour should be that you get the elements pointer and that it stays valid for as long as the `Data` or `DataConstRef` object remains valid. Or better, as long as the `Message` object remains valid, if there is one, considering that the above code would otherwise be valid by coincidence only. I think that should be possible, but I'll need to be a little bit more clever with the MsgPack decode. MsgPack is not very well documented, so this will take a bit of digging around.

Thanks for reporting!

**Comment by LourensVeen** [comment]
I've run the tests with valgrind, and it's giving some hits, so it seems there's a memory management issue there.

- [ ] Track down and fix exact problem
- [ ] Integrate valgrind into the test system so that it runs the tests with valgrind if valgrind is available
- [ ] Have the CI install valgrind

**Comment by LourensVeen** [comment]
Looks like a pretty straightforward memory leak, the pointer returned by `elements()` pointed to memory owned by a temporary, which was deconstructed before returning the pointer. The same pattern is used throughout DataConstRef, but everywhere else the data is copied out before it is returned. This is the only part where a pointer to it is returned.

I don't want the user to have to free the pointer returned by `elements()`, it should be owned by the `DataConstRef` object, so that the pointer is valid for as long as the DCR object is valid. That will require an additional pointer added to DCR, but we're not likely to have large numbers of those sitting around anyway. You'll want to copy the received data to a custom data structure, or access it dynamically, creating temporary DCRs. NRVO will probably mostly take care, so then it's only a bit of extra stack space. We'll live with that.

**Comment by LourensVeen** [comment]
Released with 0.3.2.
